File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Core/Networking/ShiftsService.swift
import Foundation

/// Service for Shifts API operations based on Memory Bank API specification
actor ShiftsService {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    // MARK: - Shifts

    /// POST /shifts/open - Открыть смену
    func openShift(planId: Int) async throws -> CurrentShift {
        let request = ShiftOpenRequest(planId: planId)
        return try await apiClient.post(path: "/shifts/open", body: request)
    }

    /// POST /shifts/close - Закрыть смену
    func closeShift(planId: Int, force: Bool = false) async throws -> CurrentShift {
        let request = ShiftCloseRequest(planId: planId, force: force)
        return try await apiClient.post(path: "/shifts/close", body: request)
    }

    /// GET /shifts/current - Получить текущую смену
    ///
    /// Может вернуть:
    /// - Смену из shifts_fact (OPENED или CLOSED)
    /// - Смену из shifts_plan (NEW)
    /// - nil (если нет смен)
    ///
    /// - Parameter assignmentId: ID назначения для синхронизации с LAMA
    func getCurrentShift(assignmentId: Int? = nil) async throws -> CurrentShift? {
        var queryItems: [URLQueryItem] = []
        if let assignmentId = assignmentId {
            queryItems.append(URLQueryItem(name: "assignment_id", value: String(assignmentId)))
        }
        return try await apiClient.get(path: "/shifts/current", queryItems: queryItems)
    }

    /// GET /shifts/{id} - Получить смену по ID
    func getShift(id: Int) async throws -> CurrentShift {
        return try await apiClient.get(path: "/shifts/\(id)")
    }

    // MARK: - Stores

    /// GET /shifts/stores - Получить список всех магазинов
    func getStores() async throws -> [ShiftStore] {
        let response: StoresListResponse = try await apiClient.get(path: "/shifts/stores")
        return response.stores
    }

    /// POST /shifts/stores - Создать магазин
    func createStore(name: String, address: String?) async throws -> ShiftStore {
        let request = StoreCreateRequest(name: name, address: address)
        return try await apiClient.post(path: "/shifts/stores", body: request)
    }

    /// GET /shifts/stores/{id} - Получить магазин по ID
    func getStore(id: Int) async throws -> ShiftStore {
        return try await apiClient.get(path: "/shifts/stores/\(id)")
    }

    /// PATCH /shifts/stores/{id} - Обновить магазин
    func updateStore(id: Int, name: String?, address: String?) async throws -> ShiftStore {
        let request = StoreUpdateRequest(name: name, address: address)
        return try await apiClient.patch(path: "/shifts/stores/\(id)", body: request)
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/main/java/com/beyondviolet/wfm/core/network/ShiftsService.kt
package com.beyondviolet.wfm.core.network

import com.beyondviolet.wfm.core.models.CurrentShift
import com.beyondviolet.wfm.core.models.ShiftCloseRequest
import com.beyondviolet.wfm.core.models.ShiftOpenRequest
import com.beyondviolet.wfm.core.models.ShiftStore
import com.beyondviolet.wfm.core.models.StoreCreateRequest
import com.beyondviolet.wfm.core.models.StoreUpdateRequest
import com.beyondviolet.wfm.core.models.StoresListResponse

/**
 * Service for Shifts API operations based on Memory Bank API specification
 */
class ShiftsService(private val apiClient: ApiClient) {

    // MARK: - Shifts

    /**
     * POST /shifts/open - Открыть смену
     */
    suspend fun openShift(planId: Int): ApiResponse<CurrentShift> {
        val request = ShiftOpenRequest(planId = planId)
        return apiClient.post("/shifts/open", request)
    }

    /**
     * POST /shifts/close - Закрыть смену
     */
    suspend fun closeShift(planId: Int, force: Boolean = false): ApiResponse<CurrentShift> {
        val request = ShiftCloseRequest(planId = planId, force = force)
        return apiClient.post("/shifts/close", request)
    }

    /**
     * GET /shifts/current - Получить текущую смену
     *
     * Может вернуть:
     * - Смену из shifts_fact (OPENED или CLOSED)
     * - Смену из shifts_plan (NEW)
     * - null (если нет смен)
     *
     * @param assignmentId ID назначения для синхронизации с LAMA
     */
    suspend fun getCurrentShift(assignmentId: Int? = null): ApiResponse<CurrentShift?> {
        val queryParams = buildMap {
            assignmentId?.let { put("assignment_id", it.toString()) }
        }
        return apiClient.get("/shifts/current", queryParams)
    }

    /**
     * GET /shifts/{id} - Получить смену по ID
     */
    suspend fun getShift(id: Int): ApiResponse<CurrentShift> {
        return apiClient.getDirect("/shifts/$id")
    }

    // MARK: - Stores

    /**
     * GET /shifts/stores - Получить список всех магазинов
     */
    suspend fun getStores(): ApiResponse<StoresListResponse> {
        return apiClient.getDirect("/shifts/stores")
    }

    /**
     * POST /shifts/stores - Создать магазин
     */
    suspend fun createStore(name: String, address: String?): ApiResponse<ShiftStore> {
        val request = StoreCreateRequest(name = name, address = address)
        return apiClient.postDirect("/shifts/stores", request)
    }

    /**
     * GET /shifts/stores/{id} - Получить магазин по ID
     */
    suspend fun getStore(id: Int): ApiResponse<ShiftStore> {
        return apiClient.getDirect("/shifts/stores/$id")
    }

    /**
     * PATCH /shifts/stores/{id} - Обновить магазин
     */
    suspend fun updateStore(id: Int, name: String?, address: String?): ApiResponse<ShiftStore> {
        val request = StoreUpdateRequest(name = name, address = address)
        return apiClient.patchDirect("/shifts/stores/$id", request)
    }
}
